traffic_ops/traffic_ops_golang/deliveryservice/servers/servers.go

```diff
@@ -263,7 +243,428 @@ func (dss *TODeliveryServiceServer) Delete(db *sqlx.DB, user auth.CurrentUser) (
+					if eType == tc.DataConflictError {
+						handleErrs(http.StatusInternalServerError, err)
+						return
+					}
+					handleErrs(http.StatusInternalServerError, err)
+					return
+				}
+				log.Errorf("received non pq error: %++v from create execution", err)
+				return
+			}
+			respServers = append(respServers, servers[i])
+			resultRows.Next()
+			i++
+			defer resultRows.Close()
+		}
+
+		err = tx.Commit()
+		if err != nil {
+			log.Errorln("Could not commit transaction: ", err)
+			return
+		}
+		resAlerts := []tc.Alert{tc.Alert{"server assignements complete", "success"}}
+		repRes := tc.DSSReplaceResponse{resAlerts, tc.DSSMapResponse{*dsId, *payload.Replace, respServers}}
+
+		// marshal the results to the response stream
+		respBts, err := json.Marshal(repRes)
+		if err != nil {
+			log.Errorln("Could not marshal the response as expected: ", err)
+			handleErrs(http.StatusInternalServerError, err)
+			return
+		}
+
+		rollbackTransaction = false
+		w.Header().Set(tc.ContentType, tc.ApplicationJson)
+		fmt.Fprintf(w, "%s", respBts)
+		return
+	}
+}
+
+type TODeliveryServiceServers tc.DeliveryServiceServers
+
+var serversRef = TODeliveryServiceServers(tc.DeliveryServiceServers{})
+
+func GetServersRef() *TODeliveryServiceServers {
+	return &serversRef
+}
+
+// api/1.1/deliveryservices/{xml_id}/servers
+func GetCreateHandler(db *sqlx.DB) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		handleErrs := tc.GetHandleErrorsFunc(w, r)
+
+		// find the delivery service Id dsId matching the xml_id
+		params, err := api.GetCombinedParams(r)
+		if err != nil {
+			log.Errorf("unable to get parameters from request: %s", err)
+			handleErrs(http.StatusInternalServerError, err)
+			return
+		}
+
+		xmlId, ok := params["xml_id"]
+		if !ok {
+			log.Errorf("unable to get xml_id parameter from request: %s", err)
+			handleErrs(http.StatusInternalServerError, err)
+			return
+		}
+
+		ctx := r.Context()
+		user, err := auth.GetCurrentUser(ctx)
+		if err != nil {
+			log.Errorf("unable to retrieve current user from context: %s", err)
+			handleErrs(http.StatusInternalServerError, err)
+			return
+		}
+
+		// if the object has tenancy enabled, check that user is able to access the tenant
+		// check user tenancy access to this resource.
+		hasAccess, err, apiStatus := tenant.HasTenant(*user, xmlId, db)
+		if !hasAccess {
+			switch apiStatus {
+			case tc.SystemError:
+				handleErrs(http.StatusInternalServerError, err)
+				return
+			case tc.DataMissingError:
+				handleErrs(http.StatusBadRequest, err)
+				return
+			case tc.ForbiddenError:
+				handleErrs(http.StatusForbidden, err)
+				return
+			default:
+				e := http.StatusInternalServerError
+				handleErrs(e, err)
+				return
+			}
+		}
+
+		row := db.QueryRow(selectDeliveryService(), xmlId)
+		var dsId int
+		row.Scan(&dsId)
+
+		// get list of server Ids to insert
+		defer r.Body.Close()
+		payload := GetServersRef()
```

Comment:
This is decoding a request into a pointer to a global variable. This isn't safe, it will decode multiple concurrent requests into the same object. 

You can fix it by changing `GetServersRef()` to `s := TODeliveryServiceServers(tc.DeliveryServiceServers{}); return &s`, so every request gets a new object.